### PR TITLE
[Feature] Added --genRedist, rewrote nimrtl/nimhcr compilation docs

### LINKED ISSUE #11587: [request] Make it easier to produce nimrtl/nimhcr, reccomend in docs setting rpath to $ORIGIN
In the docs it is shown that [you can generate nimrtl](https://nim-lang.org/docs/nimc.html#dll-generation) using the following command `nim c -d:release lib/nimrtl.nim`, but this *does not work* in any arbitrary directory - you have to have the standard library folder on hand (or a copy of nimrtl.nim in your project that you build separately, which, to be fair, is very short as it includes the whole standard library directly).

So, some suggestions:

- [ ] Make it easier to be able to just build nimrtl/nimhcr from anywhere, i.e. possibly add a new compiler switch?
- [ ] Add nimhcr build instructions to the docs
- [ ] Docs might do well to suggest setting rpath to $ORIGIN on *nix systems if nimrtl/nimhcr are not installed as system libraries (which I assume people probably aren't doing) and are to be bundled with the binary in the same folder. (This doesn't need to be done for Windows, b/c on Windows DLL loading searches the app dir by default)

### PR DESCRIPTION
**Pull request by awr1**
Resolves #11587

### CODE REVIEW

File: doc/nimc.rst [hunk 1]
```diff
@@ -326,20 +326,66 @@ They are:
...
 
-  nim c -d:release lib/nimrtl.nim
+.. code-block:: Nim
+  when defined(unix): {.passL: "-Wl,-rpath='$ORIGIN'"}
+
+``nimrtl``
+----------
+
+The ``nimrtl`` libarary contains both Nim's memory manager as well as the full
+standard libary. To generate the ``nimrtl`` library, first, create a file named
```
Comment (Jjp137) [inline doc/nimc.rst]: - line 351: "libarary" -> "library"
- line 352: "libary" -> "library"

File: doc/nimc.rst [hunk 2]
```diff
@@ -326,20 +326,66 @@ They are:
...
+``nimrtl`` as well as another library, ``nimchr``. Similarly to how ``nimrtl``
+was built, you can generate this library by create a file named
+``build_nimhcr.nim``, filling it with the following:
+
+.. code-block:: Nim
+  include nimhcr
+
+Compiling with the command::
+
+  nim c -d:release -d:createNimHcr --app:lib --out:nimrtl build_nimrtl.nim
```
Comment (Jjp137) [inline doc/nimc.rst]: `--out:nimrtl build_nimrtl.nim` should be `--out:nimhcr build_nimhcr.nim` instead.

File: doc/nimc.rst [hunk 3]
```diff
@@ -326,20 +326,66 @@ They are:
...
 
   nim c -d:useNimRtl myprog.nim
 
 **Note**: Currently the creation of ``nimrtl.dll`` with thread support has
 never been tested and is unlikely to work!
 
+``nimhcr``
+----------
+Nim's hot-code-reloading support requires the app binary to link to both
+``nimrtl`` as well as another library, ``nimchr``. Similarly to how ``nimrtl``
```
Comment (Araq) [inline doc/nimc.rst]: typo, 'nimchr'.

File: doc/nimc.rst [hunk 4]
```diff
@@ -326,20 +326,66 @@ They are:
...
+
+``nimrtl``
+----------
+
+The ``nimrtl`` library contains both Nim's memory manager as well as the full
+standard library. To generate the ``nimrtl`` library, first, create a file named
+``build_nimrtl.nim``, and fill it with the following:
+
+.. code-block:: Nim
+  include nimrtl
```
Comment (Araq) [inline doc/nimc.rst]: Huh? No, 'nimrtl' is a "main Nim" file, not an include file.

File: doc/nimc.rst [hunk 5]
```diff
@@ -326,20 +326,66 @@ They are:
...
+correctly is by including the following line in your source code:
 
-  nim c -d:release lib/nimrtl.nim
+.. code-block:: Nim
+  when defined(unix): {.passL: "-Wl,-rpath='$ORIGIN'"}
+
+``nimrtl``
+----------
+
+The ``nimrtl`` library contains both Nim's memory manager as well as the full
```
Comment (Araq) [inline doc/nimc.rst]: No, not the "full" standard library, only some commonly used stuff.

File: doc/nimc.rst (hunk 4 above)
Comment (awr1) [reply doc/nimc.rst]: I don't think this is ideal too, but the problem is this is the easiest way I can think of to generate nimrtl/hcr from anywhere without having to change the compiler. The compiler won't accept `lib/nimrtl.nim` if we're not in the stdlib directory, and it's annoying for the user to have to find it (especially if `choosenim` is being used)

Comment (Araq) [reply doc/nimc.rst]: Hmm ok.

File: doc/nimc.rst [hunk 6]
```diff
@@ -326,20 +326,67 @@ They are:
...
+
+To link against ``nimrtl`` use the command::
 
   nim c -d:useNimRtl myprog.nim
 
 **Note**: Currently the creation of ``nimrtl.dll`` with thread support has
 never been tested and is unlikely to work!
 
+``nimhcr``
+----------
```
Comment (Araq) [inline doc/nimc.rst]: We now have a separate document for HCR. Just link to it?

File: doc/nimc.rst (hunk 4 above)
Comment (awr1) [reply doc/nimc.rst]: I ended up just adding a new compilation option anyway

**Comment by awr1** [comment]
Should be okay for merge (if nobody thinks there's anything wrong)

**Comment by Araq** [comment]
Sorry, but I now think this should be a ``koch.nim`` feature, don't you agree?

**Comment by awr1** [comment]
Having `--genRedist` as a compiler option makes it easy for someone to just "generate the DLLs" wherever if they need it since they will likely have Nim on their PATH, whereas with Koch (I assume as part of the standard compiler build process) the user has to copy them out from the Nim directory (which gets more confusing if you use choosenim, since it's not immediately clear where Nim is installed)

**Comment by Araq** [comment]
Well I disagree and the implementation is hacky. This simply doesn't belong into the compiler. For example:

 ``nim c --genRedist nimhcr.nim``  # ah with --genRedist the nimhcr.nim is now relative to something special.

This belongs to ``koch.nim`` and if choosenim doesn't ship it so be it.

**Pull request closed without merging**